CHANGELOG.md

```diff
@@ -1,5 +1,32 @@
 # Changelog
 
+## 0.4.0 (TBD)
+
+#### Advice provider
+- Decorator for multiplicative inverse and iNTT over quadratic extension field for the advice tape
+- Advice provider as a trait that can be injected to program execution
+- Increased flexibility to input data to the advice provider via the created `AdviceSource`
+
+#### CLI
+- Added debug subcommand for Assembly programs with forward and backward stepping
+- Improved the execution to print the VM cycles upon termination
+
+#### Assembly
+- Added instructions: clk, ext2add, ext2sub, ext2mul, ext2div, ext2neg, ext2inv, fri_ext2fold4, hash, u32checked_popcnt, u32unchecked_popcnt
+- Added constants functionality
+- Allow empty blocks for program parse
+- Refactored `mtree_cwm` for compatibility with `mtree_set`
+
+#### Processor
+- Introduced `ProgramInfo` to bind the Kernel to a program hash
```

Comment:
I'd change the title to `Processor, Prover, Verifier` and the content to:

- [BREAKING] Refactored `execute()`, `prove()`, `verify()` functions to take `StackInputs` as one of the parameters.
- [BREAKING] Refactored `prove()` function to return `ExecutionProof` (which is a wrapper for `StarkProof`).
- [BREAKING] Refactored `verify()` function to take `ProgramInfo`, `StackInputs`, and `ExecutionProof` as parameters and return a `u32` indicating security level of the verified proof.

Reply:
Done!